Generate serialization for CVPixelBufferRef

pascoej opened the pull request:
#### 9ce92f6e45c2d139e4b665c9dd4805aada7b01e3
<pre>
Generate serialization for CVPixelBufferRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=268838">https://bugs.webkit.org/show_bug.cgi?id=268838</a>
<a href="https://rdar.apple.com/122402993">rdar://122402993</a>

Reviewed by Alex Christensen.

Use generated serializers for CVPixelBufferRef instead of handrolled.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.h: Added.
(WebKit::CoreIPCCVPixelBufferRef::CoreIPCCVPixelBufferRef):
(WebKit::CoreIPCCVPixelBufferRef::leakSendRight):
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.mm: Added.
(WebKit::CoreIPCCVPixelBufferRef::sendRightFromPixelBuffer):
(WebKit::CoreIPCCVPixelBufferRef::toCF const):
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;RetainPtr&lt;CVPixelBufferRef&gt;&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;CVPixelBufferRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(cvPixelBufferRefsEqual):
(cfHolder):
(operator==):
(TEST):

Canonical link: <a href="https://commits.webkit.org/274993@main">https://commits.webkit.org/274993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1d3a3f51147173fd7fe82fa868654409858c703

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43150 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36686 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42904 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22571 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16941 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41171 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35005 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14270 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35979 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44423 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36814 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40038 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15412 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38369 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17031 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9095 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17082 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16675 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.h:
```diff
@@ -0,0 +1,76 @@
+    {
+    }
+
+    CoreIPCCVPixelBufferRef(const CoreIPCCVPixelBufferRef& inst)
```
Copy constructor seems hopefully unnecessary.  If it is necessary, would a move constructor work instead?

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.serialization.in:
```diff
@@ -0,0 +1,30 @@
+
+webkit_platform_headers: "CoreIPCCVPixelBufferRef.h"
+[CustomHeader, RValue, WebKitPlatform] class WebKit::CoreIPCCVPixelBufferRef {
+    mach_port_t leakSendRight();
```
Does this actually leak anything?  Can't we just use the MachSendRight serialization instead of going all the way to mach_port_t?

pascoej replied:
I've been unable to get it to work with the MachSendRight serialization. Regardless of if the MachSendRight is created with create or adopt in sendRightFromPixelBuffer, the test won't be able to decode the reply.

pascoej replied:
I finally figured this out. I couldn't use the MachSendRight serialization because `SerializationTestSender::performSendWithAsyncReplyWithoutUsingIPCConnection` does not pass the "attachments" from the encoder to the decoder. I'll fix that issue in this PR and use the MachSendRight serialization.

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.h:
```diff
@@ -0,0 +1,71 @@
+
+    RetainPtr<CVPixelBufferRef> toCF() const;
+
+    MachSendRight sendRight()
```
Could you add && after sendRight() to indicate this should only be called on a CoreIPCCVPixelBufferRef r-value?

pascoej committed:
Generate serialization for CVPixelBufferRef
https://bugs.webkit.org/show_bug.cgi?id=268838
rdar://122402993

Reviewed by Alex Christensen.

Use generated serializers for CVPixelBufferRef instead of handrolled.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.h: Added.
(WebKit::CoreIPCCVPixelBufferRef::CoreIPCCVPixelBufferRef):
(WebKit::CoreIPCCVPixelBufferRef::leakSendRight):
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.mm: Added.
(WebKit::CoreIPCCVPixelBufferRef::sendRightFromPixelBuffer):
(WebKit::CoreIPCCVPixelBufferRef::toCF const):
* Source/WebKit/Shared/Cocoa/CoreIPCCVPixelBufferRef.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<RetainPtr<CVPixelBufferRef>>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<CVPixelBufferRef>>::decode): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
* Tools/TestWebKitAPI/Tests/IPC/IPCSerialization.mm:
(cvPixelBufferRefsEqual):
(cfHolder):
(operator==):
(TEST):

Canonical link: https://commits.webkit.org/274993@main
